src/events/peer_cluster.py

```diff
@@ -92,55 +101,83 @@ def _on_peer_cluster_created(self, event: RelationCreatedEvent) -> None:
+                "balancer-uris": self.charm.state.peer_cluster.balancer_uris,
+                "controller-quorum-uris": self.charm.state.peer_cluster.controller_quorum_uris,
             }
         )
 
         self.charm.on.config_changed.emit()  # ensure both broker+balancer get a changed event
 
     def _on_peer_cluster_orchestrator_changed(self, event: RelationChangedEvent) -> None:
         """Generic handler for peer-cluster-orchestrator `relation-changed` events."""
+        # TODO: `cluster_manager` check instead of runs_broker
         if (
             not self.charm.unit.is_leader()
             or not self.charm.state.runs_broker  # only broker needs handle this event
-            or "balancer"
-            not in self.charm.state.balancer.roles  # ensures secrets have set-up before writing, and only writing to balancers
+            or not any(
+                role in self.charm.state.peer_cluster.roles
+                for role in [BALANCER.value, CONTROLLER.value]
+            )  # ensures secrets have set-up before writing, and only writing to controller,balancers
         ):
             return
 
         self._default_relation_changed(event)
 
         # will no-op if relation does not exist
-        self.charm.state.balancer.update(
+        self.charm.state.peer_cluster.update(
             {
                 "roles": self.charm.state.roles,
-                "broker-username": self.charm.state.balancer.broker_username,
-                "broker-password": self.charm.state.balancer.broker_password,
-                "broker-uris": self.charm.state.balancer.broker_uris,
-                "racks": str(self.charm.state.balancer.racks),
-                "broker-capacities": json.dumps(self.charm.state.balancer.broker_capacities),
-                "zk-uris": self.charm.state.balancer.zk_uris,
-                "zk-username": self.charm.state.balancer.zk_username,
-                "zk-password": self.charm.state.balancer.zk_password,
+                "broker-username": self.charm.state.peer_cluster.broker_username,
+                "broker-password": self.charm.state.peer_cluster.broker_password,
+                "broker-uris": self.charm.state.peer_cluster.broker_uris,
+                "cluster-uuid": self.charm.state.peer_cluster.cluster_uuid,
+                "racks": str(self.charm.state.peer_cluster.racks),
+                "broker-capacities": json.dumps(self.charm.state.peer_cluster.broker_capacities),
+                "zk-uris": self.charm.state.peer_cluster.zk_uris,
+                "zk-username": self.charm.state.peer_cluster.zk_username,
+                "zk-password": self.charm.state.peer_cluster.zk_password,
             }
         )
 
         self.charm.on.config_changed.emit()  # ensure both broker+balancer get a changed event
 
+    def _on_peer_cluster_broken(self, _: RelationBrokenEvent):
+        """Handle the required logic to remove."""
+        if self.charm.state.kraft_mode is None:
```

Comment:
**todo:** Similar to the other comment, we should early-exit if the condition is NOT met, and unnest the rest.